### ISSUE TITLE
Hung builds and what to do about them

### ISSUE THREAD

**Issue by michaelpj** [nixbuild/feedback#20]
I have observed hung builds in the past and I'm currently looking at a couple right now. I'm not sure what to do about them. I can stop and restart, but I don't know if that will reliably kill the hung builds or whether I'll just end up waiting on them again.

In this case the original Nix process that started them is dead, so I can't even see the logs to tell if they're progressing, I just get `Waiting 30 seconds for an ongoing build...` endlessly.

**Comment by michaelpj** [comment]
Restarted my workflow and I'm back to `Waiting 30 seconds ...` :/

**Comment by michaelpj** [comment]
```
nixbuild.net> builds list -r
1263233 2023-03-23 12:04:35Z 29:37 1 CPU [Running]
        /nix/store/53f91k9mhb5mn1y6pvrim1y680fglzmd-cardano-ledger-api-lib-cardano-ledger-api-1.0.0.0.drv
1262731 2023-03-23 11:41:27Z 52:45 1 CPU [Running]
        /nix/store/9z7m9qsc6080ih1hr0kd7p1144hh3ymp-ouroboros-network-mock-lib-ouroboros-network-mock-0.1.0.0.drv
```
neither of them should take anything like that long!

**Comment by michaelpj** [comment]
Okay wow, they did eventually terminate. Weird.

**Comment by rickynils** [comment]
On our side it looks like the SSH session was not closed until `2023-03-23 12:58:56.503401+00` which is well after you claimed your Nix process was dead. Are you using `ControlMaster` in your SSH config, by any chance? We've seen some issues in the past were disconnects went unnoticed due to control master connection being used.

I'm still investigating why the builds were hanging, it looks like these two builds just locked up.

To summarize, we really have three problems:

* The builds stuck for some reason. Might be some issue on the builder machine or some bug in our code.
* The builds were not cancelled when your Nix process closed. It looks like we didn't detect your disconnect.
* There is no other way for you to cancel these builds. We clearly should have functionality for that in the shell. I've opened up an issue for that: https://github.com/nixbuild/feedback/issues/21

**Comment by rickynils** [comment]
@michaelpj For the build of `/nix/store/9z7m9qsc6080ih1hr0kd7p1144hh3ymp-ouroboros-network-mock-lib-ouroboros-network-mock-0.1.0.0.drv` we actually detected that it had locked up, but not until one hour after the last log output from the build. It was then restarted and ran successfully, taking less than 10 seconds.

The other build (`/nix/store/53f91k9mhb5mn1y6pvrim1y680fglzmd-cardano-ledger-api-lib-cardano-ledger-api-1.0.0.0.drv`) was started after the first build, and had not been hanging for a full hour when the SSH disconnect was noticed. I suspect that we might have detected it was hanging after an hour too. But we should have heartbeats from the builders much more often than this, so I'm puzzled we didn't detect it quicker.

**Comment by michaelpj** [comment]
> On our side it looks like the SSH session was not closed until 2023-03-23 12:58:56.503401+00 which is well after you claimed your Nix process was dead. Are you using ControlMaster in your SSH config, by any chance? We've seen some issues in the past were disconnects went unnoticed due to control master connection being used.

It was triggered from a GHA workflow using `nixbuild-action`, which does the SSH setup, so :shrug:

The only thing funny was that there was another run triggered shortly afterwards before the first one died, building the same things. That was the on that ended up waiting.

**Comment by michaelpj** [comment]
Some more stuck builds:
```
nixbuild.net> builds list -r
1264149 2023-03-23 13:18:07Z 49:36 1 CPU [Running]
        /nix/store/wp53gi8q1vgj4rfwal3lqspv2xqj5v0y-ouroboros-network-protocols-lib-ouroboros-network-protocols-0.3.0.0.drv
1264087 2023-03-23 13:17:24Z 50:17 1 CPU [Running]
        /nix/store/jrzm6j5byxsbwkm25li3g6zgz7k02935-cardano-crypto-praos-lib-cardano-crypto-praos-2.1.1.0.drv
1263988 2023-03-23 13:16:07Z 51:34 1 CPU [Running]
        /nix/store/8dk66cn3rwqcs03ayjjjww1gay28rvgd-orphans-deriving-via-lib-orphans-deriving-via-0.1.0.1.drv
1263768 2023-03-23 13:12:05Z 55:38 1 CPU [Running]
        /nix/store/a1m78p29jdqfinc1hchpm8cs52if03zx-cardano-slotting-lib-cardano-slotting-0.1.0.2.drv
```

**Comment by michaelpj** [comment]
In this case the original Nix process is still running, so they're just genuinely stuck I guess. FWIW I haven't observed these kinds of derivations getting stuck before in other contexts where we build such things.

**Comment by michaelpj** [comment]
Again looks like they are eventually hitting the 1hr timeout, getting killed and restarted, and then progressing, but it's not ideal.

**Comment by rickynils** [comment]
@michaelpj I've tracked down the reason for the 1 hour timeout, and it is a stupid one. We've mistakenly configured our builder watchdog to 1 hour. I'm fixing it now and deploying it ASAP.

We haven't figured out _why_ these lockups happen for you. Perhaps there could be some resource starvation, that only happens when large enough dependency trees is built or something. We will continue our investigation and see if we can find some pattern to these lockups. Sorry for the inconvenience!

**Comment by michaelpj** [comment]
No worries, just trying to provide enough information!

**Comment by rickynils** [comment]
@michaelpj We're grateful for all this feedback! I think that remote store building is still a quite exotic thing to do, and you are probably one of the users that is using it most. Admittedly, the support for remote store building is a (relatively speaking) recent addition to nixbuild.net, so we obviously have things to iron out.

**Comment by rickynils** [comment]
@michaelpj The fix for the 1 hour watchdog was deployed a couple of hours ago

**Comment by kokobd** [comment]
@rickynils I'm also having some hung builds which should not take that long. They were created by a remote store build. Builds without using remote store were canceled successfully, so I guess this might be related to remote store.
```
nixbuild.net> builds list -r
1276395 2023-03-26 03:41:14Z 45:31:50 2 CPUs [Running]
        /nix/store/qg5h1mi6f7w1wqrs6m70il0nm8jjq9i8-haskell-src-exts-1.23.1.drv
1276393 2023-03-26 03:41:13Z 45:31:50 8 CPUs [Running]
        /nix/store/hb4m7icya6ygs0c9xapmr66vd1hlmiwz-alex-3.2.7.1.drv
1276392 2023-03-26 03:36:44Z 45:36:19 1 CPU [Running]
        /nix/store/7nfjxacai0z1gv7050777y5f7s3d3ryd-pretty-hex-1.1.drv
1276391 2023-03-26 03:36:42Z 45:36:21 1 CPU [Running]
        /nix/store/2kcqbdshx15w1n6dz3mmadv1f13md0gi-bytestring-builder-0.10.8.2.0.drv
1276390 2023-03-26 03:36:42Z 45:36:21 1 CPU [Running]
        /nix/store/dqziyp08z82cwiy4bp8kha77fg1dnd64-fmlist-0.9.4.drv
1276388 2023-03-26 03:36:42Z 45:36:21 2 CPUs [Running]
        /nix/store/660sh043d283c5wifm4ahps6a06pg144-js-dgtable-0.5.2.drv
1276387 2023-03-26 03:36:42Z 45:36:21 1 CPU [Running]
        /nix/store/j28ly7wjsj5vbil5ccqs1vy7zyh5djyb-ed25519-0.0.5.0.drv
1276385 2023-03-26 03:36:42Z 45:36:21 1 CPU [Running]
        /nix/store/3whgvfgc21da9wy5f3x98lcxzk4yjk06-js-jquery-3.3.1.drv
1276384 2023-03-26 03:36:41Z 45:36:22 1 CPU [Running]
        /nix/store/jzf2jkaqw5wh0d3qnkwk059hrd7qfays-groups-0.5.3.drv
1276383 2023-03-26 03:36:41Z 45:36:22 1 CPU [Running]
        /nix/store/hnlyy1rahp9bdhswwfpam3v1smzhzxkl-setenv-0.1.1.3.drv
1276381 2023-03-26 03:36:41Z 45:36:22 1 CPU [Running]
        /nix/store/3a6dss4d5xs4iij41hzni6qf0xdr5isz-hscolour-1.24.4.drv
1276380 2023-03-26 03:36:41Z 45:36:22 1 CPU [Running]
        /nix/store/zq41876i6mm29srkngc4hg9blx5gbw1q-base-prelude-1.6.1.drv
1276379 2023-03-26 03:36:41Z 45:36:22 2 CPUs [Running]
        /nix/store/gj4fypnrj07sd71pah3kirl638fb7hff-network-3.1.2.7.drv
1276378 2023-03-26 03:36:41Z 45:36:22 1 CPU [Running]
        /nix/store/71rb0y84lhw9ig58dli74iccikv5f0jn-terminal-size-0.3.3.drv
1276377 2023-03-26 03:36:41Z 45:36:22 1 CPU [Running]
        /nix/store/ly4id74533z8rbs4hyrik6kxlpcw036a-some-1.0.4.1.drv
1276376 2023-03-26 03:36:41Z 45:36:22 1 CPU [Running]
        /nix/store/20pfa4cfqp8a5azfzi733q3d76ja111d-simple-reflect-0.3.3.drv
1276374 2023-03-26 03:36:41Z 45:36:22 1 CPU [Running]
        /nix/store/3hycwgvkf2fpzry9dyaz25y4232cj6fr-mmap-0.5.9.drv
1276373 2023-03-26 03:36:41Z 45:36:22 1 CPU [Running]
        /nix/store/7nn1bpwsh2q9mzsykz6f6nlykn095m16-prettyprinter-1.7.1.drv
1276372 2023-03-26 03:36:41Z 45:36:22 1 CPU [Running]
        /nix/store/mgsinbh39j0pj9v07278i4bk8fzwhmkh-unliftio-core-0.2.1.0.drv
1276371 2023-03-26 03:36:41Z 45:36:22 1 CPU [Running]
        /nix/store/dd8195982dn7fcprkbajpws2b5990w4y-Only-0.1.drv
```

**Comment by rickynils** [comment]
We actually found and fixed another issue with the builder watchdog, and also introduced some extra safety measures for killing/clearing builds in case of "catastrophic" failures in nixbuild.net (like OOM etc). You should hopefully see less of these hung builds. Still, if any build would seem stuck, #21 should be used.
